Confiant AdQuality: Add config to exclude bidders from scanning, analytics tags

IvanTsymbal opened the pull request:
Hi @bretg - current PR is an enhancement for Confiant module which provides possibility to turn off scanning for bids from specific Bidders.
Point regarding analytics is not fully clear for me yet. Once I know exactly what we need - will provide separate PR with improvements. Thanks.

garygz-confiant commented on extra/modules/confiant-ad-quality/src/main/resources/module-config/confiant-ad-quality.yaml:
```diff
@@ -17,3 +17,4 @@ hooks:
         long-interval-attempts: 336
         long-interval: 1800000
       scan-state-check-interval: 600000
+      bidders-to-exclude-from-scan:
```
do we need to provide an example
is this comma separated?

garygz-confiant replied:
or the example is here: https://github.com/prebid/prebid-server-java/pull/2670/files#diff-1bb08eaea30ab0ffe869994b44c347ca6ea1d13cf5bab5dc2a4ab0a91cbfd6faR77

IvanTsymbal replied:
Right. Example is inside **README.md** file. And separately will be added to the official Prebid docs repo.

AntoxaAntoxic commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/v1/ConfiantAdQualityBidResponsesScanHook.java:
```diff
@@ -47,9 +56,14 @@ public Future<InvocationResult<AllProcessedBidResponsesPayload>> call(
+                .collect(Collectors.groupingBy(bidderResponse -> !biddersToExcludeFromScan.contains(bidderResponse.getBidder())));
+
+        final List<BidderResponse> toScan = Optional.ofNullable(needScanMap.get(true)).orElse(List.of());
+        final List<BidderResponse> avoidScan = Optional.ofNullable(needScanMap.get(false)).orElse(List.of());
```
that looks better: `needScanMap.getOrDefault(true, Collections.emptyList())`

P.S. using `Collection.emptyList()` is preferable due to the style guide

AntoxaAntoxic commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/v1/ConfiantAdQualityBidResponsesScanHook.java:
```diff
@@ -90,8 +106,12 @@ private InvocationResult<AllProcessedBidResponsesPayload> toInvocationResult(
-                                ? AllProcessedBidResponsesPayloadImpl.of(bidsScanResult.filterValidResponses(payload.bidResponses()))
+                                ? AllProcessedBidResponsesPayloadImpl.of(
+                                    Stream.concat(bidsScanResult.filterValidResponses(scannedBidderResponses).stream(),
+                                            notScannedBidderResponses.stream()).toList())
```
`scannedBidderRepsponses` are filtered twice implicitly (in the `toAnalyticsTags` and here), I believe `bidsScanResult.filterValidResponses(scannedBidderResponses)` can be done once and the result can be passed to the `toAnalyticsTags` (and I believe it should be filtered only if the `bidsScanResult.hasIssues()` returns true)

AntoxaAntoxic commented on extra/modules/confiant-ad-quality/src/test/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/AnalyticsMapperTest.java:
```diff
@@ -0,0 +1,80 @@
+        )));
+    }
+
+    private BidderResponse getBidderResponse(String bidderName, String impId, String bidId) {
```
can be static?

AntoxaAntoxic commented on extra/modules/confiant-ad-quality/src/test/java/org/prebid/server/hooks/modules/com/confiant/adquality/v1/ConfiantAdQualityBidResponsesScanHookTest.java:
```diff
@@ -142,6 +164,65 @@ public void shouldSubmitBidsToScan() {
+                .call(allProcessedBidResponsesPayload, auctionInvocationContext);
+
+        // then
+        verify(bidsScanner, times(1)).submitBids(
```
times(1) is the default

AntoxaAntoxic commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/AnalyticsMapper.java:
```diff
@@ -0,0 +1,81 @@
+
+    private static List<Result> toActivityResults(BidsScanResult bidsScanResult,
+                                           List<BidderResponse> scannedBidderResponses,
+                                           List<BidderResponse> notScannedBidderResponses) {
```
please format

AntoxaAntoxic commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/v1/ConfiantAdQualityBidResponsesScanHook.java:
```diff
@@ -94,6 +95,22 @@ private InvocationResult<AllProcessedBidResponsesPayload> toInvocationResult(
+                } else {
+                    bidderResponsesWithoutIssues.add(scannedBidderResponses.get(i));
+                }
+            }
```
I'd say that the `bidsScanResult` should take responsibility of this piece of code, the only difference with the `filterValidResponses` method is this code returns both valid and invalid responses. In that way your can optimize the code even further because currently `hasIssues` iterates over all the responses also

IvanTsymbal replied:
Good point. Thanks for a "side" view. Going to apply logic update.

IvanTsymbal replied:
Wasn't sure about proper name. Hope current one is quite clear and updates can resolve your idea about optimisation.

IvanTsymbal commented:
@bretg fyi - I'm gonna create related PR in the documentation repository once current is finalized and merged.

AntoxaAntoxic commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/BidsScanResult.java:
```diff
@@ -25,11 +27,24 @@ public boolean hasIssues() {
-                .filter(ind -> !hasIssuesByBidIndex(ind))
-                .mapToObj(responses::get)
-                .toList();
+    public Map<Boolean, List<BidderResponse>> toIssuesExistencyMap(List<BidderResponse> bidderResponses) {
```
I'd suggest creating a separate return object for responses (with and without issues) in order to make it super clear, the method name could just `groupByIssues`

Another question: do the `public boolean hasIssues()` is still used?

IvanTsymbal replied:
Thanks again. `groupByIssues` is already inside updates. 
Regarding `hasIssues` I had the same question to myself. But wanted to keep it for the future development and some tests scenarious. But if you have such question as well - means would be better to clean it now and add once we need that. So removed.

AntoxaAntoxic reviewed (dismissed):
@SerhiiNahornyi I've finished with the review and approved, please proceed

IvanTsymbal commented:
> @bretg fyi - I'm gonna create related PR in the documentation repository once current is finalized and merged.

Related PR for documentation repo: [feature: Add ExcludeBidders from scan config and Analytic tags for Confiant Ad Quality module docs](https://github.com/prebid/prebid.github.io/pull/4949)
Preview: [Module docs](https://deploy-preview-4949--prebid-docs-preview.netlify.app/prebid-server/pbs-modules/confiant-ad-quality)

IvanTsymbal commented:
@And1sS  hello. I have fixed code style changes which you've mentioned, removed redundant OperationalResult model and moved Object mapper away from method to the phase of initialisation.

But I did not touch **toGroupByIssues**. Because of the Analytics features I need to have 2 different lists of bids - with issues and without issues. So to avoid 2 iterations to filter bids with 2 different aims - current mapping gives advantages. 

Will it be a problem to leave it as is? If so - i can split it into separate class helper which can do such mapping. What would you say?

Thanks.

bretg commented:
Thanks @IvanTsymbal - please update the docs -- https://github.com/prebid/prebid.github.io/blob/master/prebid-server/pbs-modules/confiant-ad-quality.md

IvanTsymbal commented:
@bretg I've already mentioned that related doc updates are there. So hope everything is fine:

Related PR for documentation repo: https://github.com/prebid/prebid.github.io/pull/4949
Preview: [Module docs](https://deploy-preview-4949--prebid-docs-preview.netlify.app/prebid-server/pbs-modules/confiant-ad-quality)
